pageserver/src/tenant/layer_map.rs

```diff
@@ -733,14 +733,25 @@ where
 
     #[inline(always)]
     fn compare_arced_layers(left: &Arc<L>, right: &Arc<L>) -> bool {
-        // FIXME: ptr_eq might fail to return true for 'dyn' references because of multiple vtables
-        // can be created in compilation. Clippy complains about this. In practice it seems to
-        // work.
+        // "dyn Trait" objects are "fat pointers" in that they have two components:
+        // - pointer to the object
+        // - pointer to the vtable
         //
-        // In future rust versions this might become Arc::as_ptr(left) as *const () ==
-        // Arc::as_ptr(right) as *const (), we could change to that before.
-        #[allow(clippy::vtable_address_comparisons)]
-        Arc::ptr_eq(left, right)
+        // rust does not provide a guarantee that these vtable's are unique, but however
+        // `Arc::ptr_eq` in 1.66.1 uses a comparison where both the pointer and the vtable need to
+        // be equal.
+        //
+        // See: https://github.com/rust-lang/rust/issues/103763
+        //
+        // A future version of rust will most likely use this form below, where we cast each
```

Comment:
Sorry, I do not understand this comment and why we need to case pointer to "const*?
Why it is not possible just to do Arc::as_ptr(left) == Arc::as_ptr(right) as suggested in Rust doc:
```
/// Provides a raw pointer to the data.
    ///
    /// The counts are not affected in any way and the `Arc` is not consumed. The pointer is valid for
    /// as long as there are strong counts in the `Arc`.
    ///
    /// # Examples
    ///
    /// ```
    /// use std::sync::Arc;
    ///
    /// let x = Arc::new("hello".to_owned());
    /// let y = Arc::clone(&x);
    /// let x_ptr = Arc::as_ptr(&x);
    /// assert_eq!(x_ptr, Arc::as_ptr(&y));
    /// assert_eq!(unsafe { &*x_ptr }, "hello");
    /// ```
    #[must_use]
    #[stable(feature = "rc_as_ptr", since = "1.45.0")]
    pub fn as_ptr(this: &Self) -> *const T {
        let ptr: *mut ArcInner<T> = NonNull::as_ptr(this.ptr);

        // SAFETY: This cannot go through Deref::deref or RcBoxPtr::inner because
        // this is required to retain raw/mut provenance such that e.g. `get_mut` can
        // write through the pointer after the Rc is recovered through `from_raw`.
        unsafe { ptr::addr_of_mut!((*ptr).data) }
    }
```

Reply:
I'll try to come up with playground link to illustrate this.

Reply:
Does this illustrate the two-pointer nature of `*const dyn Trait` as given by `Arc::as_ptr` in this case?

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0d1038e802b245e2f29d9264b8b077d8

Reply:
Also compare to `Arc::ptr_eq` impl: https://doc.rust-lang.org/stable/src/alloc/sync.rs.html#1150

So this is what we are hoping to do differently, until std adds a different method or changes that `Arc::ptr_eq` method, as has been voted on the linked issue.

Reply:
Also compare to `Arc::ptr_eq` impl: https://doc.rust-lang.org/stable/src/alloc/sync.rs.html#1150

So this is what we are hoping to do differently, until std adds a different method or changes that `Arc::ptr_eq` method, as has been voted on the linked issue.